### PR TITLE
Multiple chat improvements

### PR DESCRIPTION
**Pull request by wladekb**
This PR combines changes from two former PRs - https://github.com/Wikia/app/pull/10215 and https://github.com/Wikia/app/pull/10213

This changeset aims to fix the following issues:
- https://wikia-inc.atlassian.net/browse/SUS-422
- https://wikia-inc.atlassian.net/browse/SUS-360
- https://wikia-inc.atlassian.net/browse/SUS-361
- https://wikia-inc.atlassian.net/browse/SUS-336
- https://wikia-inc.atlassian.net/browse/SUS-360
- https://wikia-inc.atlassian.net/browse/SUS-416
- https://wikia-inc.atlassian.net/browse/SUS-417
- https://wikia-inc.atlassian.net/browse/SUS-379
- https://wikia-inc.atlassian.net/browse/SUS-358
- https://wikia-inc.atlassian.net/browse/SUS-335

possibly:
- https://wikia-inc.atlassian.net/browse/SUS-365 (hard to reproduce on devbox)

Summary of changes/fixes includes:
- major overhaul of PHP and JS code structure
  - rename ChatEntryPoint to ChatWidet
  - create a model-like ChatUser class
  - fix naming of many variables and methods
  - remove unused code - unused methods and entire failover functionality
- get rid of ChatConfig.json and use a few variables and consul instead
- improve performance of DOM operations in ChatWidget.js
- redo widget templates using mustache
- fix display issues with chat users carousel in widget
- tests

Depends on:
- Wikia/chat#25
- Wikia/config#1738

Feel free to add comments and suggestions!

/cc @dianafa @owend @SeanColombo @Wikia/sustaining-engineering-team @sactage

### CODE REVIEW

File: extensions/wikia/Chat2/ChatAjax.class.php [hunk 1]
```diff
@@ -208,63 +229,91 @@ static public function getListOfBlockedPrivate() {
...
+		$tmpl = new EasyTemplate( __DIR__ . '/templates/' );
+		$tmpl->set_vars( [
+			'options' => Chat::getBanOptions(),
+			'isChangeBan' => $isChangeBan,
+			'isoTime' => $isoTime,
+			'fmtTime' => $fmtTime
+		] );
+
+		$res = [
+			'template' => $tmpl->render( "banModal" ),
```
Comment (kerriamber) [inline extensions/wikia/Chat2/ChatAjax.class.php]: Should it be `'banModal'` (single quotes)?

File: includes/GlobalFunctions.php [hunk 2]
```diff
@@ -3824,8 +3824,9 @@ function wfWaitForSlaves( $wiki = false ) {
 			$lb = clone $lb;
 
 			for ( $i=0; $i < count( $slaves ); $i++ ) {
+				list( $slaveHost, $slavePort ) = explode( ':', $slaves[$i] );
```
Comment (dianafa) [inline includes/GlobalFunctions.php:3827]: can we fix spacing in this whole method?

File: extensions/wikia/Chat2/tests/ChatUserTest.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
+<?php
+
+class ChatUserText extends WikiaBaseTest {
+
+	const MOCK_USER_ID = 142;
+	const MOCK_WIKI_ID = 189;
+	const MOCK_ADMIN_USER_ID = 289;
```
Comment (dianafa) [inline extensions/wikia/Chat2/tests/ChatUserTest.php:7]: any particular reason to choose this numbers?

File: extensions/wikia/Chat2/js/controllers/controllers.js [hunk 4]
```diff
@@ -91,7 +91,8 @@ var NodeChatSocketWrapper = $.createClass(Observable, {
 
 		};
 
-		this.proxy(callback, this)('name=' + encodedWgUserName + '&key=' + wgChatKey + '&roomId=' + this.roomId + '&serverId=' + this.serverId);
+		this.proxy(callback, this)('name=' + encodedWgUserName + '&key=' + wgChatKey + '&roomId=' + this.roomId
```
Comment (dianafa) [inline extensions/wikia/Chat2/js/controllers/controllers.js:89]: TODO: add comment above to remember to remove this param: https://wikia-inc.atlassian.net/browse/SUS-432

File: extensions/wikia/Chat2/ChatUser.class.php [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+	private function getBanInfoFromCache( &$existed ) {
+		$banInfo = $this->wg->Memc->get( $this->getBanInfoCacheKey() );
+
+		$existed = !empty( $banInfo );
+		if ( $banInfo === self::NO_BAN_MARKER ) {
+			$banInfo = false;
+		}
+
+		return $banInfo;
```
Comment (dianafa) [inline extensions/wikia/Chat2/ChatUser.class.php]: what about `return $banInfo === self::NO_BAN_MARKER ? false : $banInfo;` ?

File: extensions/wikia/Chat2/ChatConfig.class.php [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		return $serverNodes[$index];
+	}
+
+	public static function getSecretToken() {
+		global $wgChatCommunicationToken;
+
+		return $wgChatCommunicationToken;
+	}
+
+}
```
Comment (dianafa) [inline extensions/wikia/Chat2/ChatConfig.class.php]: newline

File: extensions/wikia/Chat2/ChatBanTimeOptions.class.php [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	 * @return ChatBanTimeOptions
+	 */
+	public static function newDefault() {
+		$textSource = wfMessage( 'chat-ban-option-list' )->inContentLanguage()->text();
+		$textSource = preg_replace( '!\s+!', ' ', $textSource );
+
+		return new self( $textSource );
+	}
+
+}
```
Comment (dianafa) [inline extensions/wikia/Chat2/ChatBanTimeOptions.class.php]: newline

File: extensions/wikia/Chat2/templates/widget.mustache [hunk 8]
```diff
@@ -23,37 +23,7 @@
...
-												<span class="icon">&nbsp;</span>
-												<span class="label" data-msg-id="chat-user-menu-contribs"> </span>
-											</a>
-										</li>
-									</ul>
-								</div>
-							</div>
-						</li>
-						{{/users}}
+						{{> widgetUserElement}}
```
Comment (wladekb) [inline extensions/wikia/Chat2/templates/widget.mustache]: Just saying the space here should be optional.

**Pull request merged**